File: Data Structures & Algorithms/sort-an-array/submission-0.cpp
class Solution {
public:
    vector<int> num;
    vector<int> sortArray(vector<int>& nums) {
        num = nums;
        sort(0, num.size()); 
        return num;
    }

    void sort(int start, int end)
    {
        if(start>=end-1)
        {
            return;
        }
        int i = start-1;
        for(int j = start; j<end-1; j++)
        {
            if(num[j]<=num[end-1])
            {
                i++;
                int temp = num[i];
                num[i] = num[j];
                num[j] = temp;
            }
        }
        int temp = num[i+1];
        num[i+1] = num[end-1];
        num[end-1] = temp;
        sort(start, i+1);
        sort(i+2,end);
    }
};

File: Data Structures & Algorithms/target-sum/submission-0.java
class Solution {
    int[][] dp;
    int targ;
    int maxx;
    int[] num;
    public int findTargetSumWays(int[] nums, int target) {
        num = nums;
        targ = target;
        int max = 0;
        for(int k = 0; k<nums.length; k++)
        {
            max+=nums[k];
        }
        maxx = max;
        dp = new int[nums.length][2*max+1];
        for(int i = 0; i<dp.length; i++)
        {
            for(int j = 0; j<dp[0].length; j++)
            {
                dp[i][j] = -1;
            }
        }
        return recurse(0, 0);
    }

    public int recurse(int index, int sum)
    {
        if(index==num.length)
        {
            if(sum==targ)
            {
                return 1;
            }
            return 0;
        }
        if(dp[index][sum+maxx]!=-1)
        {
            return dp[index][sum+maxx];
        }
        int results = recurse(index+1, sum+num[index]) + recurse(index+1, sum-num[index]);
        dp[index][sum+maxx] = results;
        return dp[index][sum+maxx];
    }
}
